rs/index/src/segment/pending_segment.rs

```diff
@@ -93,8 +94,43 @@ impl<Q: Quantizer + Clone + Send + Sync> PendingSegment<Q> {
 
     // Caller must hold the write lock before calling this function.
     pub fn apply_pending_deletions(&self) -> Result<()> {
-        // TODO(hicder): Implement this once we support deletions.
-        Ok(())
+        let internal_index = self.index.read();
+        match &*internal_index {
+            Some(index) => {
+                let invalidated_ids_directory =
+                    PathBuf::from(format!("{}/invalidated_ids_storage", index.base_directory));
+                if !invalidated_ids_directory.exists() {
+                    return Err(anyhow!("Invalidated ids directory does not exist"));
+                }
+                if !invalidated_ids_directory.is_dir() {
+                    return Err(anyhow!("Invalidated ids path is not a directory"));
+                }
+
+                // At this point there should be no invalidated ids recorded in internal index.
+                let is_empty = fs::read_dir(&invalidated_ids_directory)?.next().is_none();
+                if !is_empty {
+                    return Err(anyhow!("Invalidated ids directory is not empty"));
+                }
+
+                // TODO(tyb): hard link the storage? But still need to invalidate in the hash set
+                for (user_id, doc_ids) in self.temp_invalidated_ids.read().iter() {
+                    for doc_id in doc_ids {
+                        if !index.invalidate(*user_id, *doc_id)? {
+                            return Err(anyhow!(
```

Comment:
what happens if `doc_id` just doesn't exist in `index`?

Reply:
Right, not a big deal if doc_id doesn't exist in index, we should just skip it. Updated the code